File: scripts/display_log.c
/* For the size of the file. */
#include <sys/stat.h>
#include <sys/mman.h> 
#include <errno.h>
#include <string.h>
#include <stdarg.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>

/* logging */
typedef struct {
	unsigned long  modified;     /* seconds since epoch*/
	unsigned char  protocol;     /* values from: prot_t*/
	unsigned char  action;       /* valid values: NF_ACCEPT, NF_DROP*/
	unsigned char  hooknum;      /* as received from netfilter hook*/
	unsigned int   src_ip;
	unsigned int   dst_ip;
	unsigned short src_port;
	unsigned short dst_port;
	signed int     reason;       /* rule#, or values from: reason_t*/
	unsigned int   count;        /* counts this line's hits*/
} log_row_t;

static void
check (int test, const char * message, ...)
{
    if (test) {
        va_list args;
        va_start (args, message);
        vfprintf (stderr, message, args);
        va_end (args);
        fprintf (stderr, "\n");
        exit (EXIT_FAILURE);
    }
}

int main ()
{
    /* The file descriptor. */
    int fd;
    /* Information about the file. */
    struct stat s;
    int status;
    size_t size;
    /* The file name to open. */
    const char * file_name = "/dev/fw5_log";
    /* The memory-mapped thing itself. */
    const char * mapped;
    int i;
    log_row_t* lg;

    /* Open the file for reading. */
    fd = open (file_name, O_RDONLY);
    check (fd < 0, "open %s failed: %s", file_name, strerror (errno));

    /* Get the size of the file. */
    status = fstat (fd, & s);
    check (status < 0, "stat %s failed: %s", file_name, strerror (errno));
    /*size = s.st_size;*/
    size = 28672;

    /* Memory-map the file. */
    mapped = mmap (0, size, PROT_READ, MAP_SHARED, fd, 0);
    check (mapped == MAP_FAILED, "mmap %s failed: %s",
           file_name, strerror (errno));

    /* Now do something with the information. */
    lg = (log_row_t *)mapped;
    for (i = 0; lg->protocol && (((void*)lg) - ((void*)mapped)) < size; i++) {
	printf("%d: mod: %lu prt: %u act: %u hoo: %u sa: %u da: %u\n"
		"\t\tsp: %u dp: %u rs: %u ct: %u\n", i,
		lg->modified, lg->protocol, lg->action,
		lg->hooknum, lg->src_ip, lg->dst_ip, lg->src_port, 
		lg->dst_port, lg->reason, lg->count);
	++lg;
    }

    return 0;
}

File: scripts/display_log.py
#!/usr/bin/python3

import mmap, time, ast, struct, socket

structs = {
    'log': 'I 3B 2I 2H i I',
    'rule': '>4B 2H 2I',
    'conn': '2I 2H I B 3x',
}            

class Log:
    def __init__(self, l):
        if not isinstance(l, str):
            raise Exception('bad log init: l = {}'.format(l))
	#print "len: " + repr(len(l)) + " calcsize: " + repr(struct.calcsize(structs['log']))
	if len(l) != struct.calcsize(structs['log']):
            raise Exception('bad log init: l = {} calcsize'.format(l))

        (self.time, self.prot, self.action, self.hooknum,
            self.src_ip, self.dst_ip,
            self.src_port, self.dst_port,
            self.reason, self.count) = struct.unpack(structs['log'], l)

with open('/dev/fw5_log', 'rb') as log, \
     open('/sys/class/fw5/fw5_log/log_size') as log_size_file:
     log_size = ast.literal_eval(log_size_file.readline())
     l = mmap.mmap(log.fileno(), log_size, mmap.MAP_SHARED, mmap.PROT_READ)
     i = 0
     log = Log(l[i * 28: (i + 1) * 28])
     print "log_size: " + repr(log_size) + " prot: " + repr(ord(l[i * 28 + 4])) + "\n"
     while ord(l[i * 28 + 4]):
         log = Log(l[i * 28: (i + 1) * 28])
         i += 1
         print repr(i) + "# tm: " + repr(log.time) + " pr: " + repr(log.prot) + " ac: " + repr(log.action) + " hk: " + repr(log.hooknum) + " sa: " + socket.inet_ntoa(struct.pack('L',socket.ntohl(log.src_ip))) + " sp: " + repr(log.src_port) + " da: " + socket.inet_ntoa(struct.pack('L',socket.ntohl(log.dst_ip))) + " dp: " + repr(log.dst_port) + " rs: " + repr(log.reason) + " ct: " + repr(log.count)

     l.close
